src/lib_json/json_writer.cpp

```diff
@@ -1217,9 +1215,7 @@ bool StreamWriterBuilder::validate(Json::Value* invalid) const {
   std::set<String> valid_keys;
   getValidWriterKeys(&valid_keys);
   Value::Members keys = settings_.getMemberNames();
-  size_t n = keys.size();
-  for (size_t i = 0; i < n; ++i) {
-    String const& key = keys[i];
+  for (const auto& key : keys) {
```

Comment:
`getMemberNames()` is inefficient.
We should be able to do with without a `keys` vector.
Iterate directly through `settings_`, copying elements into `inv`.

Reply:
I tried this patch:

```
--- a/src/lib_json/json_writer.cpp
+++ b/src/lib_json/json_writer.cpp
@@ -1212,12 +1212,7 @@ bool StreamWriterBuilder::validate(Json::Value* invalid) const {
   Json::Value& inv = *invalid;
   std::set<String> valid_keys;
   getValidWriterKeys(&valid_keys);
-  Value::Members keys = settings_.getMemberNames();
-  for (const auto& key : keys) {
-    if (valid_keys.find(key) == valid_keys.end()) {
-      inv[key] = settings_[key];
-    }
-  }
+  std::copy_if(settings_.begin(), settings_.end(), inv.begin(), [=](const Json::Value& key){return valid_keys.find(key) == valid_keys.end();});
   return inv.empty();
 }
 Value& StreamWriterBuilder::operator[](const String& key) {
```

Not working.

Reply:
You can't do it with iterator-hiding algorithms, because *iterator doesn't expose object keys. You also can't insert at Value.begin() like that. It is not an insert_iterator.

Please pull this change for a working refactor of the validate function:

https://github.com/BillyDonahue/jsoncpp/commit/3b416b03ab533fa6c2418e2d0c8223031206ac49